### ISSUE TITLE
JQUERY UPGRADE

### ISSUE THREAD

**Issue by aethelwulffe** [LibreHealthIO/LibreEHR#416]
My understanding of JQUERY:

-  Version 1.x supports ancient I.E.
-  Version 2.x has the same API as 1.x, while dropping the old IE compatibility.
-  Version 3.x works a little differently.
-  JQuery-Compat 3.x takes over the job of the 1.x for compatibility of IE8, Opera12, and Safari5 and other junk.
-  JQuery 3.x does not have the compatibility bits. 
- Both 3.x varieties use the same public API, and otherwise match features by version. 

https://github.com/jquery/jquery-migrate/ is a tool we can use to identify and migrate jquery versions. In my opinion, we should be able to test everything with this, upgrade code where needed and all that.

Tasks:

- [x] Place the package from https://code.jquery.com/jquery-3.1.1.min.js into the assets/js directory.
- [x] Add the jquery_current directory to globals.
- [x] Identify all uses of the deprecated jquery functions listed below.
- [x] Modify all instances of deprecated API calls to use the latest versions.
- [x] Create and retain a maintenance/change log for JQUERY.

Check off deprecated functions as they are searched for and upgraded.

1.3
- [x] jQuery.boxModel
- [x] jQuery.browser

 1.7
- [x] deferred.isRejected()
- [x] deferred.isResolved()
- [x] .die()
- [x] jQuery.sub()
- [x] .live()
- [x] .selector

 1.8
- [x] .andSelf()
- [x] deferred.pipe()
- [x] .error()
- [x] .load()
- [x] .size()
- [ ] .toggle()
- [x] .unload()

 1.9
- [x] jQuery.support
 
 1.10
- [x] .context

 3.0
- [x] .bind()
- [x] .delegate()
- [x] jQuery.fx.interval
- [x] .unbind()
- [x] .undelegate()

**Comment by teryhill** [comment]
I sent up 3.1.1 and the migrate tool. Do you want to send the ones that are currently used in portal and upgrade them later?

**Comment by teryhill** [comment]
Any particular global name you want ?

**Comment by aethelwulffe** [comment]
No opinion.  Make it unambiguous. (means "talk plain").

**Comment by teryhill** [comment]
so calling it $globals['the stuff that does the cool stuff] is probably out huh?

$GLOBALS['js_path']
$GLOBALS['images_path']
$GLOBALS['css_path']
$GLOBALS['fonts_path']

**Comment by aethelwulffe** [comment]
No usage of calls like this->Thingie and stuff.
I think a "jquery_currentver or the like might be useful.  Anyone else on that?

**Comment by teryhill** [comment]
$GLOBALS['js_path'] = "$web_root/assets/js"; then we could put the current version in a directory called current_version and keep the ones that are needed as jquery-2.2.0.min etc

**Comment by aethelwulffe** [comment]
.selector appears to used in the zend module care coordination.  It has it's own copy of jq though.
The TAGS code has TWO copies of the jquery migration package in it.
.andSelf() is in the "custom" 1.8.21 versions.  Obviously "Custom" means "retain deprecated functions".
require.js in /tags_filters has a lot of .load() calls, I think so if you are using a Playstation 3 everything still works...  this is a resource though, and I don't see it used in the code.

Care Coordination uses .toggle()
summary_pat_portal.php uses .toggle()
The default ajax templates for day, week month like:
month/ajax_template.html uses .toggle()
patient_report.php uses toggle()

**Comment by teryhill** [comment]
summary_pat_portal.php uses .toggle() is going away with the new portal so mark it off the list. since the tag's has it's own assets directory I vote we leave it and let @kchapple make the changes if he feels they are needed. lets just concentrate on the core items.

**Comment by kchapple** [comment]
Thanks terry. Great work. 

Ken
kenchapple.com
facebook.com/kenchapplemusic
@ken_chapple

Check out my latest CD, Crossing the Black at https://itun.es/us/SGGI9

**Comment by teryhill** [comment]
ready for the Yelling?

**Comment by teryhill** [comment]
I am going to start pushing up the portal assets and then we can sort out what we need.

**Comment by aethelwulffe** [comment]
Terry, the calendar still uses .toggle(), so unless that is in ajax and is a different function/library completely, we still have an issue.

**Comment by teryhill** [comment]
Lets just put that version of jquery in the assets and wait till google fixes it 8^)

**Comment by aethelwulffe** [comment]
patient_report.php uses .toggle() with phimail to send CCR and CCD.  It looks like the biggest concern is in the versions of ajax_template.html under PostCalendar.

**Comment by aethelwulffe** [comment]
Yeah, by all means, we get the new version of JQuery in, and a global pushed for assets.  I'm just sweating the actual migration at this point.

**Comment by aethelwulffe** [comment]
#415 merged up.
It's Machete Time again!   ..now where is that bar oil and the _fine tooth_ chainsaw?

**Comment by teryhill** [comment]
just hack the jquery's for right now. I will do another commit to get the rest of the portal assets up there and then I will make the changes to point it to the assets and then we repeat the process on the portal I know a couple version of jquery are used.

**Comment by aethelwulffe** [comment]
They might just be 1. and 2.x, which is just for compatibility sake.  If 
that is so, and they don't have the deprecated functions, and there are 
no custom version and function calls, then we would be OK with 3.x

**Comment by teryhill** [comment]
From what I read about that migrate tool it should fix the problems as long as they aren't fatal.

**Comment by aethelwulffe** [comment]
Yeah, sure, but I haven't figured anything else about it...like how the heck to use it.  Is it just adding:
`<script src="webroot/assets/js/jquery-3.x.x.js"></script>
<script src="webroot/assets/js/jquery-migrate-3.x.x.js"></script>`
...and then go to town?

**Comment by teryhill** [comment]
That's what I got

**Comment by teryhill** [comment]
I see that they have a 1.4.1 version to get to 1.9 +
<script src="http://code.jquery.com/jquery-1.11.3.js"></script>
<script src="http://code.jquery.com/jquery-migrate-1.4.1.js"></script>

**Comment by aethelwulffe** [comment]
I think you only have to be on version 1.3 or higher for the 3.0 to work.  That is what the docs say anyway...I think.

**Comment by aethelwulffe** [comment]
Hope we can get back to this pretty soon!

**Comment by aethelwulffe** [comment]
#561 is a move intended to promote migrating the js includes.

**Comment by teryhill** [comment]
I am working on 1.3.2 clean out. Moving to 1.7.2 or greater where possible.

**Comment by aethelwulffe** [comment]
#569 Moving most of the datatables package stuff over.  There is some...delicate...stuff in /tags that I am leery to touch at the moment that seems like assets that could be merged with this package, and there are css bits as well, but I am just leaving most of the structure be right now.

**Comment by aethelwulffe** [comment]
This is all checked off now.  Took four months, but I am pretty sure we can just call the latest JQUERY anywhere we need.
We now have a scheme and an institutional awareness of using the assets directory and header loading include.  Still have lots of stray js packages all over the place, but this specific issue is done.

**Comment by nileshprasad137** [comment]
@aethelwulffe Yep. There are many places where jquery 1.x are used which were using` .load()` which got removed in 3.0. We will need to replace that with `.on("load",function(){})`

**Comment by aethelwulffe** [comment]
Crap.  Sorry I closed this.  I thought I searched for "load(" and came up blank.  I must have not formed my search parameters correctly.  Sublimetext has a good search history though, so I can see where I screwed up.
